modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java

```diff
@@ -124,14 +124,20 @@ Message dequeue() {
        
         Message message = queue.dequeue();
         if (message != null) {
-            pendingMessages.put(message.getMetadata().getInternalId(), message);
+            pendingMessages.put(message.getMetadata().getInternalId(), message.shallowCopy());
         }
         return message;
     }
 
     void acknowledge(long messageId) {
         // TODO handle nacks
-        pendingMessages.remove(messageId);
+        Message message = pendingMessages.remove(messageId);
+        message.release();
+    }
+
+    public void requeue(long messageId) {
+        Message message = pendingMessages.remove(messageId);
+        queue.enqueue(message);
```

Comment:
So we are putting the message into end of the queue? So it get delivered much later. Specification doesn't say anything on this right?

Reply:
Noted. Created https://github.com/wso2/message-broker/issues/93 to track this